src/shims/unix/linux/epoll.rs

```diff
@@ -95,42 +148,92 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        let ready_list = &epoll_file_description.ready_list;
+
+        let Some(file_descriptor) = this.machine.fds.dup(fd) else {
+            drop(epfd);
+            return Ok(Scalar::from_i32(this.fd_not_found()?));
+        };
+        let weak_file_descriptor = file_descriptor.get_weak_file_descriptor();
 
         if op == epoll_ctl_add || op == epoll_ctl_mod {
-            let event = this.deref_pointer_as(event, this.libc_ty_layout("epoll_event"))?;
+            // Read event bitmask and data from epoll_event struct.
+            let events = this.read_scalar(&this.project_field(&event, 0)?)?.to_u32()?;
+            let data = this.read_scalar(&this.project_field(&event, 1)?)?.to_u64()?;
 
-            let events = this.project_field(&event, 0)?;
-            let events = this.read_scalar(&events)?.to_u32()?;
-            let data = this.project_field(&event, 1)?;
-            let data = this.read_scalar(&data)?;
-            let event = EpollEvent { events, data };
+            // We only support edge-triggered notification for now.
+            if events & epollet != epollet {
+                throw_unsup_format!("epoll_ctl: epollet flag must be included.");
+            }
 
-            let Some(mut epfd) = this.machine.fds.get_mut(epfd) else {
-                return Ok(Scalar::from_i32(this.fd_not_found()?));
+            let epoll_key = (weak_file_descriptor, fd);
+
+            // Check if the fd is already in the interest list.
+            if op == epoll_ctl_add {
+                if interest_list.contains_key(&epoll_key) {
+                    let eexist = this.eval_libc("EEXIST");
+                    drop(epfd);
+                    this.set_last_error(eexist)?;
+                    return Ok(Scalar::from_i32(-1));
+                }
+            } else {
+                if !interest_list.contains_key(&epoll_key) {
+                    let enoent = this.eval_libc("ENOENT");
+                    drop(epfd);
+                    this.set_last_error(enoent)?;
+                    return Ok(Scalar::from_i32(-1));
+                }
+            }
+
+            let weak_file_descriptor = file_descriptor.get_weak_file_descriptor();
+            let binding = weak_file_descriptor.get_file_description().unwrap();
+            let target_file_description = binding.borrow_mut();
+            let event = EpollEvent {
+                file_descriptor: fd,
+                weak_file_descriptor,
+                events,
+                data,
+                ready_list: Rc::clone(ready_list),
             };
-            let epfd = epfd
-                .downcast_mut::<Epoll>()
-                .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_ctl`"))?;
+            // Modify the interest list entry if it already exists (epoll_ctl),
+            // and insert if it doesn't (epoll_add).
+            interest_list.entry(epoll_key).and_modify(|e| *e = event.clone()).or_insert(event);
```

Comment:
If you are overwriting the entire value, you can just insert instead of using the entry api. Inserting overwrites previous values

Reply:
Right, I noticed the implementation here is wrong too, because both the data and event bitmask of epoll_event can be updated.